rancher/rancher
# Support image digest in helm chart

So that Rancher chart can be deployed in public cloud such as Azure[1].

[1] https://learn.microsoft.com/en-us/partner-center/marketplace/azure-container-technical-assets-kubernetes?tabs=linux#update-the-helm-chart

## Issue: https://github.com/rancher/rancher/issues/43061
 
## Problem
Microsoft Azure Marketplace container offer required all the relevant helm charts to use image hash instead of label. See https://learn.microsoft.com/en-us/partner-center/marketplace/azure-container-technical-assets-kubernetes?tabs=linux#update-the-helm-chart

"Images must use digests instead of tags. This ensures CNAB building is deterministic."

Therefore, we cannot create a container offer by using the upstream Rancher helm chart as is right now. We work around it by keeping a local modified copy. It would be nice for upstream to add this feature.
 
## Solution
Introduce an extra (optional) value `rancherImageDigest` in `values.yaml`. `rancherImageDigest` and `rancherImageTag` should be mutually exclusive. `rancherImageDigest` should take precedent over `rancherImageTag` if both are present in `values.yaml`.
 
## Testing

## Engineering Testing
### Manual Testing
The following scenarios had been manually tested.

1. default
2. explicitly specifying `rancherImageTag` to make sure it's being used properly
3. explicitly specifying `rancherImageDigest` to make sure it's being used properly
4. explicitly specifying both `rancherImageDigest` and `rancherImageTag` to make sure only `rancherImageDigest` is being used.

### Automated Testing
None. Not sure if there's existing automation tests for the charts.

Summary: _TODO_

## QA Testing Considerations
Make sure the above scenarios are covered.
 
### Regressions Considerations
Make sure the default scenario is covered.

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_
